pkg/hub/managedclusterset/controller.go

```diff
@@ -0,0 +1,265 @@
+package managedclusterset
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+
+	clientset "github.com/open-cluster-management/api/client/cluster/clientset/versioned"
+	clusterinformerv1 "github.com/open-cluster-management/api/client/cluster/informers/externalversions/cluster/v1"
+	clusterinformerv1alpha1 "github.com/open-cluster-management/api/client/cluster/informers/externalversions/cluster/v1alpha1"
+	clusterlisterv1 "github.com/open-cluster-management/api/client/cluster/listers/cluster/v1"
+	clusterlisterv1alpha1 "github.com/open-cluster-management/api/client/cluster/listers/cluster/v1alpha1"
+	clusterv1alpha1 "github.com/open-cluster-management/api/cluster/v1alpha1"
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	operatorhelpers "github.com/openshift/library-go/pkg/operator/v1helpers"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog"
+)
+
+// managedClusterSetController reconciles instances of ManagedClusterSet on the hub.
+type managedClusterSetController struct {
+	clusterClient    clientset.Interface
+	clusterLister    clusterlisterv1.ManagedClusterLister
+	clusterSetLister clusterlisterv1alpha1.ManagedClusterSetLister
+	eventRecorder    events.Recorder
+}
+
+// NewManagedClusterSetController creates a new managed cluster set controller
+func NewManagedClusterSetController(
+	clusterClient clientset.Interface,
+	clusterInformer clusterinformerv1.ManagedClusterInformer,
+	clusterSetInformer clusterinformerv1alpha1.ManagedClusterSetInformer,
+	recorder events.Recorder) factory.Controller {
+	c := &managedClusterSetController{
+		clusterClient:    clusterClient,
+		clusterLister:    clusterInformer.Lister(),
+		clusterSetLister: clusterSetInformer.Lister(),
+		eventRecorder:    recorder.WithComponentSuffix("managed-cluster-set-controller"),
+	}
+	return factory.New().
+		WithInformersQueueKeyFunc(func(obj runtime.Object) string {
+			accessor, _ := meta.Accessor(obj)
+			return accessor.GetName()
+		}, clusterSetInformer.Informer()).
+		WithInformers(clusterInformer.Informer()).
+		WithSync(c.sync).
+		ToController("ManagedClusterSetController", recorder)
+}
+
+func (c *managedClusterSetController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	key := syncCtx.QueueKey()
+
+	// if the sync is triggered by change of ManagedCluster, reconcile all cluster sets
+	if key == "key" {
+		if err := c.syncAllClusterSets(ctx); err != nil {
```

Comment:
make a TODO to better than this.  You only need to recheck the ones that contain the managed cluster in question, right?  You can build this with a reverse index.

Reply:
> make a TODO to better than this. You only need to recheck the ones that contain the managed cluster in question, right? You can build this with a reverse index.

Even if you have to list everything, the workqueue behavior will help reduce storms if you compute the individual names.

Reply:
As currently written, with sufficient number of clusters this implementation will not finish reconciling before the refresh interval is exhausted.

Reply:
TODO is added and will deal with it in a follow-up PR.